Gpg additions master

github-abcde opened the pull request:
### What does this PR do?
It adds a gpg state module, and cleans up the gpg execution module bigtime. It also adds loads of unit- and integration tests.
Things cleaned up in the gpg execution module:
- Made function return values consistently return a dict (where appropriate).
- Add all parameters in function docstrings and ensure proper python Sphinx domain formatting.
- Replace incorrect GPG_1_3_1 version check with version checks that distinguish between the version of python's gnupg library and the version of the installed GPG tools.
- Add `gnupghome`-parameter to all relevant functions.
- Adds passphrase usage options, for example when creating/importing/exporting secret keys with GPG 2.1 or later, a passphrase is now required.
- Improved passphrase-from-pillar support. Now no longer requires a fixed pillar key `gpg_passphrase`, but the pillar key can now be passed as argument `passphrase_pillar`.
- Add signing/verifying with detached signatures.
- Change default keyserver to use `pool.sks-keyservers.net`.
- Add symmetric encrypting/decrypting support.
- A lot of errorhandling has been added.

Changes to the gpg state module:
- Updated function docstrings to contain all parameters.
- Added lots of extra tests.
- All state functions are now truly idempotent (except when idempotency cannot be determined, for example when encrypting data, it cannot be determined that existing encrypted data is the same as newly encrypted data due to the random factors used in encryption).
- Added `trusted` state to ensure keys exist at a certain trust level.
- Added `data_encrypted`, `data_decrypted`, `data_signed` and `data_verified` state functions.

### What issues does this PR fix or reference?
None that I am aware of

### Previous Behavior
A lot of functionality is missing.

### New Behavior
Data encryption/decryption/signing/verification has been extended and made more robust.
State module has been extended with encryption/decryption/signing/verification states.

### Tests written?
Verily

### Commits signed with GPG?
Yes

github-abcde committed:
WIP: Mostly rewrite. Change function results to return more consistent data.
Add encryption/decryption/signing/verification functions.
Update function docstrings.

github-abcde committed:
WIP: Almost complete rewrite.
Added function docstrings.
Updated present/absent to handle changes in returned datastructure in modules/gpg.py.
Added functions for encryption/decryption/signing/verification.

github-abcde committed:
WIP: Rewritten and expanded tests.

github-abcde committed:
WIP: Add integration tests

github-abcde committed:
Fix unittests

github-abcde committed:
Apply autoformatting

github-abcde committed:
Remove unneeded newline

github-abcde committed:
Add more tests

github-abcde committed:
Unittest-inspired improvements

github-abcde committed:
Add more tests

github-abcde committed:
Fix export tests

github-abcde committed:
Fix pylint-directive being ignored

github-abcde committed:
Add integration test_gpg test 10,11

github-abcde committed:
Move pylint comment back to its original position so it will work
Make text a mandatory argument to the text_encrypt function.

github-abcde committed:
Add missing CheckError import

github-abcde committed:
Remove unused imports

github-abcde committed:
Remove hardcoded hashing algorithm from tests

github-abcde commented:
It seems the testsuite defaults to using `python-gnupg` version `0.4.4` which has some bugs that cause the `integration.states.test_gpg` test to fail. Is it possible to start using version `0.4.5`?

github-abcde committed:
Fix use of @depends decorator
cannot compare None with >=

github-abcde committed:
Fix doc build

github-abcde committed:
Remove whitespace to fix linting error

github-abcde committed:
Bump version of python-gnupg to 0.4.5

github-abcde committed:
Change GPG_VERSION check to work with both tests and doc

github-abcde committed:
Replace custom cleanup with unittest-provided cleanup
Replace custom tempdir with RUNTIME_VARS.TMP

github-abcde committed:
Replace custom tempdir with RUNTIME_VARS.TMP

github-abcde committed:
Replace import of salt.modules.gpg with check for functions in __salt__

github-abcde committed:
Removed unused import

github-abcde committed:
Apply yapf styling

twangboy commented:
@dmurphy18 Could you review this. It's quite a large change and I seem to recall you doing something with gpg recently.

twangboy commented:
@garethgreenaway Would you mind taking a look. Looks like you wrote most of the original module.

dmurphy18 reviewed (commented):
The codecov report platforms tested on does not include latest Ubuntu or Debian.
Would like to see this tested against Ubuntu 18.04, Debian 10 and Centos 8, since these have later versions of gnupg and gpg-agent functionality which has given issues in the past.

github-abcde commented:
@dmurphy18 Can you give me any specific things to test? (other than the things already tested in the integration tests).

dmurphy18 commented:
@github-abcde Wanted to make sure that the tests have been run on the platforms listed above is all, before I started reviewing, since that are later platforms with different OS version of gnupg installed

github-abcde committed:
Ignore missing files during cleanup
Fix test_get_fingerprint_from_data to handle different results from GPG versions

github-abcde committed:
Perform dry-run import with temp homedir
> This fixes cases where the user salt is running under does not have an accessible homedir

github-abcde committed:
Fix incorrect handling of already imported key.

github-abcde committed:
Update signed_data mock with realistic data
Remove test that did not test any code functionality in salt
Force GPG_VERSION to 2.1 for test_create_key_fail without password

github-abcde committed:
Account for duplicate secret keys in import_key
Include GPG_VERSION > 2.0 requirement for --pinentry-mode argument
Check res.status more specifically when signing

github-abcde committed:
Distinguish between GPG v1 and GPG v2 returns
Always include passphrase when needed (even though GPG v2 caches this between tests)

github-abcde commented:
I'll debug centos7 next week :)

github-abcde committed:
Change distinction in GPG behavior between old/new
I was wrong to make the distinction between v1 and v2. Instead it had to be
between everything up to and including v2.0 and v2.1+
Also fix linting errors.

github-abcde committed:
Change approach to ignore absent files during rm_rf

github-abcde committed:
Fix incorrect raise statement

github-abcde commented:
@dmurphy18 The tests currently failing are not caused by my code changes, so it's all yours if you want to have a look :-)

github-abcde committed:
Applied black formatting

github-abcde committed:
Apply isort

github-abcde committed:
Apply black

twangboy commented:
@github-abcde Hey, would you mind rebasing and fixing the conflicts here?

github-abcde committed:
Add slowtest-decorators as per PR56774
Any test that takes longer than 1s to run should have this decorator.

github-abcde commented:
@twangboy Done

github-abcde commented:
@sagetherage Aww, why was this postponed?

sagetherage commented:
it still needs to be reviewed and hasn't passed all checks prior to code freeze

github-abcde commented:
@sagetherage The failing checks all fail on parts outside the changes I have made. They have for some time (at least since Mar 17). Rebasing (or rather, pressing the button [Update branch] is trivial. I'm not sure if I'm the only one able to do that, but I'd rather press the button once (and start another full cycle of tests) shortly before it gets merged, than having to press the button every time the master branch gets updated (and start a cycle of tests that nobody will use).

github-abcde committed:
Merge branch 'master' into gpg_additions-master

garethgreenaway commented:
@github-abcde Looks good, once you fix the requests from Shane and those lint failures.

github-abcde commented:
@sagetherage Unfortunately, I can't fix the lint failures, as they are caused by black. If I fix them, the test that checks for changes black would cause would fail.
For example: https://github.com/saltstack/salt/pull/56283/commits/29ad6568a924a85be7cf5a09557db87cfbd2ee75#diff-b8b7c8d8ad2997fb6d709835e153d8e5R908

sagetherage commented:
@github-abcde ok we'll look into the lint failures

garethgreenaway commented:
@github-abcde For the lint issues just remove the commas at the end of the lines where the lint failures are being triggered.  That will resolve the issues.  I was about to fix it myself but it looks like your repo doesn't allow us to edit your pull requests.

github-abcde committed:
Fix linting errors

github-abcde committed:
Implement review suggestions
Restored the availability of the "use_passphrase"-argument to the functions that had them.
Added deprecation message until version Phosphorus.

github-abcde committed:
Merge branch 'master' into gpg_additions-master

github-abcde committed:
Fix silly typo

github-abcde committed:
Merge branch 'master' into gpg_additions-master

github-abcde committed:
Apply black formatting

github-abcde commented:
@twangboy Can you resolve the points you opened if you deem them sufficiently handled?

waynew commented on doc/conf.py:
```diff
@@ -148,22 +148,23 @@ def inner(fn, *iargs, **ikwargs):  # pylint: disable=unused-argument
+    "ansible",
     "avahi",
     "boto.regioninfo",
+    "concurrent",
```
Looks like this is causing the docs failure

bryceml replied:
this is fixed on master I think.  This probably needs to be rebased and will then be fine.

github-abcde replied:
Doesn't look like it. On the other hand, it looks like I added `ansible` and `concurrent` to fix the doc build on my end, but apparently broke it in master :)

bryceml replied:
yup, that sounds correct after looking at it closer.

github-abcde committed:
Merge branch 'master' into gpg_additions-master

github-abcde committed:
doc/conf.py: Remove concurrent from mock list as this breaks the build on Jenkins

github-abcde commented:
@dwoz Failing docs test has been resolved.

github-abcde committed:
Merge branch 'master' into gpg_additions-master

github-abcde committed:
Merge branch 'master' into gpg_additions-master

github-abcde committed:
Merge branch 'master' into gpg_additions-master

github-abcde committed:
Merge branch 'master' into gpg_additions-master

github-abcde committed:
Merge branch 'master' into gpg_additions-master

github-abcde committed:
Merge branch 'master' into gpg_additions-master

github-abcde committed:
Merge branch 'master' into gpg_additions-master

github-abcde committed:
Fix pre-commit test

github-abcde committed:
Merge branch 'master' into gpg_additions-master

sagetherage commented:
@github-abcde Windows tests are failing, so we are looking into that, ponce done we will re-run these

github-abcde commented:
@sagetherage Thanks! I tried to reproduce it on my windows10 installation, but the tests do not fail there.
I briefly tried to figure out what's going on, and I suspect it tries to install a specific version of 7zip that is no longer the `latest` version. Whether that's actually what's failing the test I don't know though :-). Regardless, I could not find a connection to my changes in the gpg module and the failing windows test (and since a few days also the failing test on centos7).

sagetherage commented:
@github-abcde yes, it is our tests failing, I looked this yesterday and meant to put the comment in, apologies I didn't, but yes, I will get back to this once those tests are fixed on our end :)

github-abcde committed:
Merge branch 'master' into gpg_additions-master

github-abcde committed:
Apply black 20.8b1

github-abcde committed:
Merge branch 'master' into gpg_additions-master

sagetherage commented:
looks like this needs some attention and I will have to remove it from the Magnesium release cycle

twangboy commented:
This needs another rebase...

twangboy commented on salt/states/gpg.py:
```diff
@@ -32,149 +23,841 @@
+                    or data.startswith(r"\x85\x01\x0c\x03")  # 1024b RSA encrypted data
+                    or data.startswith(r"\x85\x01\x8c\x03")  # 2048b RSA encrypted data
+                    or data.startswith(r"\x85\x02\x0c\x03")  # 3072b RSA encrypted data
+                    or data.startswith(  # 3072b RSA encrypted data
```
The comment here looks strange

dmurphy18 reviewed (changes requested):
@github-abcde Are you still interested in this PR ?, if unresponsive the PR shall be closed after several days.
Test cases need addressing as well as review comments, and rebasing given requirements have changed since this PR was written
Also don't notice a changelog entry

github-abcde commented:
@dmurphy18 I am!
However I am disappointed in the way this MR has been handled from the side of the saltstack maintainers. In all the time this MR is open, it has mostly been waiting on input from your end, and having to rebase things _yet again_ is not the most pleasant of activities.
I will see when I can find time to go through this again.

dmurphy18 commented:
@github-abcde Pointing out that there were requested changes
```
[twangboy](https://github.com/twangboy) requested changes on behalf of https://github.com/orgs/saltstack/teams/team-core [on Feb 11, 2021](https://github.com/saltstack/salt/pull/56283#pullrequestreview-588624135)
```
And there are conflicts which must be resolved before continuing, perhaps you failed to notice those.

Apologize for any delays on Salt's behalf in reminding you of those issues.

dwoz commented:
Closing this due to inactivity. Anyone should feel free to re-open it if they want to see it through to the end in one release cycle.
